pkg/kallsyms/kallsyms_test.go

```diff
@@ -77,78 +84,169 @@ func TestRealKAllSyms(t *testing.T) {
+	require.ErrorContainsf(t, err, "was not found in kallsyms", "specUpdateAddresses should have failed")
 
 	err = specUpdateAddresses(
-		kAllSymsFactory,
+		[]symbolResolver{
+			kSymbolResolver,
+		},
 		spec,
 		[]string{"bpf_prog_fops", "socket_file_ops"},
 	)
-	require.Nil(t, err, "specRewriteConstantsWithSymbolAddresses failed: %v", err)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
 
 	expectedContents := []byte{}
 	expectedContents = append(expectedContents, bpfProgFopsAddr...)
 	expectedContents = append(expectedContents, socketFileOpsAddr...)
 	contents := spec.Maps[".rodata"].Contents[0].Value
 	require.Equal(t, contents, expectedContents, "contents aren't equal")
 }
+
+func TestSpecRewriteConstantsRoot(t *testing.T) {
+	utilstest.RequireRoot(t)
+	utilstest.RequireFileContains(t, "/proc/kallsyms", "bpf_prog_fops")
+	utilstest.RequireFileContains(t, "/proc/kallsyms", "socket_file_ops")
+	t.Cleanup(resetState)
+
+	resetState()
+
+	spec1 := specTest.Copy()
+	err := specUpdateAddresses(
+		[]symbolResolver{
+			newKAllSymsResolver(),
+		},
+		spec1,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
+
+	resetState()
+
+	spec2 := specTest.Copy()
+	err = specUpdateAddresses(
+		[]symbolResolver{
+			newEbpfResolver(),
+		},
+		spec2,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
+
+	resetState()
+
+	spec3 := specTest.Copy()
+	err = specUpdateAddresses(
+		[]symbolResolver{
+			newKAllSymsResolver(),
+			newEbpfResolver(),
+		},
+		spec3,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
+
+	resetState()
+
+	spec4 := specTest.Copy()
+	err = specUpdateAddresses(
+		[]symbolResolver{},
+		spec4,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.ErrorIs(t, err, os.ErrNotExist, "specUpdateAddresses should have failed")
+
+	t.Logf("spec1: %v", spec1.Maps[".rodata"].Contents[0].Value)
+	t.Logf("spec2: %v", spec2.Maps[".rodata"].Contents[0].Value)
+	t.Logf("spec3: %v", spec3.Maps[".rodata"].Contents[0].Value)
+
+	require.Equal(t,
+		spec1.Maps[".rodata"].Contents[0].Value,
+		spec3.Maps[".rodata"].Contents[0].Value,
+		"contents spec1 and spec3 aren't equal")
+	require.Equal(t,
+		spec1.Maps[".rodata"].Contents[0].Value,
+		spec2.Maps[".rodata"].Contents[0].Value,
+		"contents spec1 and spec2 aren't equal")
+}
+
+func TestSpecRewriteConstantsWithEbpf(t *testing.T) {
+	utilstest.RequireRoot(t)
+	t.Cleanup(resetState)
+
+	spec := specTest.Copy()
+	err := specUpdateAddresses(
+		[]symbolResolver{
+			newEbpfResolver(),
+		},
+		spec,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "SpecUpdateAddresses failed: %v", err)
+	require.NotZero(t, spec.Maps[".rodata"].Contents[0].Value, "Contents should not be zero")
```

Comment:
`spec.Maps[".rodata"].Contents[0]..Value` is a `[]byte` and its length should be 16 (2 pointers) with two non-zero values. I think `require.NotZero` does not test that.

https://github.com/stretchr/testify/blob/edd74b24a1751990eeb831ebc70902c36600c3af/assert/assertions.go#L1682

I think `reflect.Zero()` for a `[]byte` would be an empty slice, so the test `require.NotZero` would always pass, even if it is a `[]byte` of size 16 with only zeros.

Reply:
How about this?

```
func TestSpecRewriteConstantsWithEbpf(t *testing.T) {
	utilstest.RequireRoot(t)
	t.Cleanup(resetState)

	spec := specTest.Copy()
	err := specUpdateAddresses(
		[]symbolResolver{
			newEbpfResolver(),
		},
		spec,
		[]string{"bpf_prog_fops", "socket_file_ops"},
	)
	require.Nil(t, err, "SpecUpdateAddresses failed: %v", err)
	require.Len(t, spec.Maps[".rodata"].Contents[0].Value, 16, "Contents should have 16 bytes")
	values, ok := spec.Maps[".rodata"].Contents[0].Value.([]byte)
	require.True(t, ok, "Contents should be a byte slice")
	emptyBytes := []byte{0, 0, 0, 0, 0, 0, 0, 0}
	require.NotEqual(t, emptyBytes, values[:8], "First byte should not be zero")
	require.NotEqual(t, emptyBytes, values[8:], "Last 8 bytes should be zero")
}
```